_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+
+```javascript
+function base64urlToUint8array(base64Bytes) {
+    const padding = '===='.substring(0, (4 - (base64Bytes.length % 4)) % 4);
+    return base64js.toByteArray((base64Bytes + padding).replace(/\//g, "_").replace(/\+/g, "-"));
+}
+
+function uint8arrayToBase64url(bytes) {
+    if (bytes instanceof Uint8Array) {
+        return base64js.fromByteArray(bytes).replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
+    } else {
+        return uint8arrayToBase64url(new Uint8Array(bytes));
+    }
+}
+```
+
+#### Authentication ceremony
+
+Registering a credential is the first step in the authentication ceremony. Using a credential is similar, but the JSON objects provided to and produced by the client are slightly different. Additionally, the function the client calls is `navigator.credentials.get()` instead of `create()`.
+
+The client provides a username, which is sent to the server using a `fetch` request. This triggers the creation of an `AssertionRequest` request object, which includes an associated `publicKey`, a list of allowed `credentialIds`, and a generated `challenge` (creating the requirement that the `AssertionRequest` is cached server-side for security).
+
+This `AssertionRequest` is passed to the client in a JSON response body. Any byte fields are again converted into Uint8Array objects, and the object is passed into `navigator.credentials.get()`. The resulting `AuthenticatorAssertionResponse` is Base64Url encoded and sent to the server as a JSON string. `clientDataJson`, like the server origin and challenge, are checked for tampering against the cached `AssertionRequest`; `authenticatorData` like `signatureCount` is updated and saved; and, most importantly, the `signature` field is checked against the stored `publicKey` generated during the creation ceremony. If all of these checks pass, the user is assigned a session, successfully signed in, and granted access to the application.
+
+{% img blog/webauthn-java/login-page.png alt:"Sign-in with authentication frontend" width:"800" %}{: .center-image }
+
+### Testing the application
+
+To test this application, pull out any [CTAP2-compliant authentication device](https://fidoalliance.org/certification/fido-certified-products/). The easiest is activating WebAuthn emulation in [Chrome](https://developer.chrome.com/docs/devtools/webauthn/). Using one of these devices, open the network inspector and watch the network requests to get a sense of the data passed back and forth between the client and server.
+
+There are other devices that can be used for WebAuthn. For a complete list, you can do a search for [FIDO-certified products](https://fidoalliance.org/certification/fido-certified-products/):
+
+* Chromebook PIN
+* Windows Hello
+* YubiKey
+* Apple Touch ID
+
+{% img blog/webauthn-java/app-welcome-page.png alt:"Successful authentication" width:"800" %}{: .center-image }
```

Comment:
It seems like this could use something like instructions to start the Spring Boot app and open a browser at 8080, etc. Or maybe just something that says you should see this image after authenticating.